group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -1336,6 +1422,233 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
+            ).orElse(defaultAssignor.name());
+            try {
+                TargetAssignmentBuilder assignmentResultBuilder =
+                    new TargetAssignmentBuilder(groupId, groupEpoch, assignors.get(preferredServerAssignor))
+                        .withMembers(group.members())
+                        .withStaticMembers(group.staticMembers())
+                        .withSubscriptionMetadata(subscriptionMetadata)
+                        .withTargetAssignment(group.targetAssignment())
+                        .addOrUpdateMember(memberId, updatedMember);
+                TargetAssignmentBuilder.TargetAssignmentResult assignmentResult;
+                // A new static member is replacing an older one with the same subscriptions.
+                // We just need to remove the older member and add the newer one. The new member should
+                // reuse the target assignment of the older member.
+                if (staticMemberReplaced) {
+                    assignmentResult = assignmentResultBuilder
+                        .removeMember(member.memberId())
+                        .build();
+                } else {
+                    assignmentResult = assignmentResultBuilder
+                        .build();
+                }
+
+                log.info("[GroupId {}] Computed a new target assignment for epoch {} with '{}' assignor: {}.",
+                    groupId, groupEpoch, preferredServerAssignor, assignmentResult.targetAssignment());
+
+                records.addAll(assignmentResult.records());
+                targetAssignment = assignmentResult.targetAssignment().get(memberId);
+                targetAssignmentEpoch = groupEpoch;
+            } catch (PartitionAssignorException ex) {
+                String msg = String.format("Failed to compute a new target assignment for epoch %d: %s",
+                    groupEpoch, ex.getMessage());
+                log.error("[GroupId {}] {}.", groupId, msg);
+                throw new UnknownServerException(msg, ex);
+            }
+        }
+
+        // 3. Reconcile the member's assignment with the target assignment if the member is not
+        // fully reconciled yet.
+        updatedMember = maybeReconcile(
+            groupId,
+            updatedMember,
+            group::currentPartitionEpoch,
+            targetAssignmentEpoch,
+            targetAssignment,
+            ownedTopicPartitions,
+            records
+        );
+
+        if (newMemberCreated) {
+            scheduleConsumerGroupSessionTimeout(groupId, memberId);
+        }
+
+        // TODO: what will happen to the client if the generation id in response is 0?
+        responseFuture.complete(new JoinGroupResponseData()
+            .setMemberId(updatedMember.memberId())
+            .setGenerationId(updatedMember.memberEpoch())
+            .setProtocolType(ConsumerProtocol.PROTOCOL_TYPE)
+            .setProtocolName(protocols.stream().findAny().get().name())
+        );
```

Comment:
This is incorrect. The issue is that it would send the response before the records are persisted/committed. I wonder if we could rather set the response to the `CoordinatorResult` below. Have you considered this?

On a similar line of thoughts, I wonder if we could rationalize the flows in this method. We sometimes completes the response future, or we throw exceptions, or we complete the response future with an error. I wonder if we could for instance use exceptions for errors and the CoordinatorResult for the responses. I am not sure whether it is actually possible.

Reply:
The responseFuture is set like this because we need to comply with `GroupCoordinatorService#joinGroup` where `responseFuture` is returned instead of the result of `scheduleWriteOperation`. To resolve `send the response before the records are persisted/committed`, we should either 1) know the group type in GroupCoordinatorService to decide whether to directly return the result of `scheduleWriteOperation`  or complete the future only when the records have been persisted

For the errors, since we need to set the member id in the response for some errors and not for the others, the handling can be a little messy. Maybe we can unify them to all throw exceptions when handling errors.

Reply:
Yeah, I understand this but I don't like it. I think that we can keep it as it is in order to minimize the changes in this PR. We can save this for later.

That being said, my initial comment still stands. You cannot complete the future immediately here. You need to use a "write future" and complete the "response future" only when the write is completed.